### PR TITLE
refactor: introduce lint tree as single discovery source

### PR DESCRIPTION
**Pull request by not-stbenjam**
## Summary

- Introduce **lint tree** — a polymorphic tree data structure that represents everything lintable in a repository
- `LintTarget` base class with `walk()`, `find()`, `content_blocks()` for tree traversal
- `PluginNode` / `SkillNode` typed markers for structural rules to discover via `tree.find(PluginNode)`
- `ContentBlock` becomes ABC with `FileContentBlock`, `FrontmatterContentBlock`, `CodeRabbitContentBlock` subclasses
- Each subclass encapsulates its format logic (`read_body()`, `write_body()`, `gather()`)
- `build_lint_tree()` replaces fragmented discovery — single source of truth
- `context.lint_tree` lazy property with `rebuild_lint_tree()` for invalidation during fix loops
- Unified fix pipeline: all fixes go through block tools, no separate file-level fix path
- Group violations by block identity (`__eq__`/`__hash__`), not `id()` or file path
- Fix double-offset bug in detectors (9 sites) — detectors now store body-relative line numbers
- Fix multi-block relint filtering with `v.block == block` instead of file path matching
- Add `skillsaw tree` subcommand to visualize the lint tree
- Design doc at `docs/designs/lint-tree.md`

## Test plan

- [x] `make test` — 1044 passed, 14 skipped
- [x] `make format` / `make lint` — clean
- [x] `make update` — regenerated docs
- [x] Test against `openshift-eng/ai-helpers` — exit 0
- [x] `skillsaw tree` displays correct tree structure
- [ ] CI passes
- [ ] CodeRabbit review

🤖 Generated with [Claude Code](https://claude.com/claude-code)

## Summary by CodeRabbit

* **New Features**
  * Added a CLI command to display the repository lint tree and discovered content blocks.
  * LLM autofix gains block-scoped edit tools for targeted, re-linted fixes.

* **Improvements**
  * Introduced a content-block model for more accurate body reading, line mapping, and round-trip edits.
  * Formatter outputs (text/HTML/JSON/SARIF) now report file-mapped line numbers.

* **Tests**
  * Expanded tests covering lint-tree, content-block read/write, coderabbit round-trips, and LLM block tools.

* **Documentation**
  * New design doc and README updates describing the lint tree.

* **Chores**
  * Package/version metadata updated.

### CODE REVIEW

**Comment by not-stbenjam** [comment]
@coderabbitai Please review this PR, focusing on the overall architecture of the ContentBlock abstraction, the block tools design, and the LLM fix pipeline refactoring. Key areas to evaluate:

1. Is the ContentBlock abstraction (with read_body/write_body/_writer) well-designed for extensibility to future composite file types?
2. Are the pathless block tools (read_block, write_block, etc.) a good API for LLM-assisted fixing?
3. Is the dual-path fix pipeline (block-based vs file-based) clean and maintainable?
4. Does the backward compatibility approach (ContentFile alias, file_path derived from block) minimize breakage risk?

**Comment by not-stbenjam** [comment]
## Review Panel Verdict

**Disposition**: APPROVE

---

### Specialist Findings

**Architecture Reviewer**

The ContentBlock refactor is well-structured and follows the existing codebase architecture cleanly.

- `NOTE` **Clean abstraction**: ContentBlock encapsulates path, category, line offset, body, custom writer, and custom line map into a single dataclass. The `file_line()` method provides a clean translation layer between block-local and file-global line numbers.
- `NOTE` **Backward compat aliases preserved**: `ContentFile = ContentBlock` and `gather_all_content_files = gather_all_content_blocks` maintain backward compatibility as required by project rules.
- `SUGGESTION` **`_make_coderabbit_writer` YAML path parsing**: The bracket-based path traversal logic (lines 237-255 of `content_analysis.py`) uses a hand-rolled path parser with `split(".")` and manual `[` bracket extraction. This works for the current set of CodeRabbit YAML paths but could break for keys containing dots or brackets. The input is controlled (generated internally by `_extract_instructions`), so this is safe in practice but worth noting for future maintainers.
- `SUGGESTION` **Block rollback with shared file paths** (`linter.py:811-835`): When multiple ContentBlocks share the same file (e.g., multiple instruction sections in `.coderabbit.yaml`), `_relint_file` filters violations only by resolved file path, not by block identity. After fixing block A, violations from block B could inflate the `after_count`, causing unnecessary (but safe) rollbacks. A future improvement could carry block references through the relint to enable per-block filtering.
- `NOTE` **id(block) for identity**: Using `id()` for block identity in the `block_violations` dict is safe here because all block objects are kept alive in `block_map` for the duration of `llm_fix()`, preventing address reuse.

No blocking issues found. Module boundaries are clean, the linter pipeline (context -> config -> rule -> linter) is preserved, and new code follows existing patterns.

**Python Expert**

- `NOTE` **Local annotations using 3.10+ syntax** (`content_analysis.py:43-44`): `list[str]` and `str | None` are used as local variable annotations in a function body. These are safe on Python 3.9+ because PEP 526 specifies that local variable annotations in function bodies are not evaluated at runtime. However, for style consistency with the rest of the codebase (which uses `List[str]` and `Optional[str]` from `typing`), consider using the uppercase forms. Not blocking.
- `NOTE` **Type hints**: All new public interfaces (`ContentBlock`, `RuleViolation.file_line`, block tools) have appropriate type hints. The `_writer` and `_line_map` callbacks use `Callable` correctly.
- `NOTE` **FileCache implementation** (`utils.py`): The thread-safe `FileCache` with per-file invalidation is a solid replacement for `lru_cache`. The two-level dict design (`resolved_path -> {sub_key -> value}`) enables O(1) per-file invalidation without disturbing other threads. The eviction strategy (drop half when full) is reasonable for a linter workload.
- `NOTE` **Dataclass usage**: `ContentBlock` properly uses `field(default=None, repr=False)` for internal callbacks, keeping repr output clean.

No blocking issues found. The code is idiomatic, well-typed, and compatible with the declared Python 3.9+ requirement.

**Security & Supply Chain Reviewer**

- `NOTE` **Path traversal protection**: `_resolve_safe()` in `llm/tools.py` correctly prevents path traversal by checking `is_relative_to(root.resolve())`. This is applied consistently across all file-based tools.
- `NOTE` **New dependency `ruamel.yaml>=0.18`**: This is a well-maintained YAML library (the reference YAML parser for Python). It's used for round-trip YAML parsing to preserve comments and formatting when writing back to `.coderabbit.yaml`. The dependency is justified and actively maintained.
- `NOTE` **LiteLLM as optional dependency**: Properly isolated behind `[llm]` extra and lazy-imported in `_litellm.py`. The `_get_litellm()` helper raises a clear error if not installed.
- `NOTE` **No subprocess usage**: Confirmed no `subprocess`, `os.system`, `eval`, or `exec` calls in any changed files.
- `NOTE` **`_make_coderabbit_writer` writes to disk**: The writer writes back to the same `.coderabbit.yaml` file that was read. The YAML path is internally generated (not user-supplied), so there's no injection risk.

No security concerns found.

**QA Engineer**

- `NOTE` **Extensive test coverage**: 1023 tests pass. The PR adds tests for all new components:
  - `TestContentBlockReadWrite` — `read_body()`, `write_body()`, `file_line()` with offset and custom line map
  - `TestBlockState`, `TestReadBlockTool`, `TestWriteBlockTool`, `TestReplaceBlockSectionTool`, `TestDiffBlockTool`, `TestLintBlockTool` — full coverage of block tools
  - `TestGatherAllContentFiles` — content block gathering with APM, excludes, deduplication
  - Content rule tests, CodeRabbit rule tests, APM rule tests, action/review.py crash-path tests
- `SUGGESTION` **Missing test: multiple blocks from same file in LLM fix rollback**: The rollback logic for blocks sharing a file path (described in Architecture review) lacks a test case. A test with a `.coderabbit.yaml` containing 2+ instruction blocks where one is fixed and one isn't would validate the rollback behavior. This is a quality gap but not blocking since the current behavior (unnecessary rollback) is the safe default.
- `SUGGESTION` **Missing test: `_make_coderabbit_writer` with bracket paths**: The writer's YAML path traversal for `path_instructions[pattern]` entries could benefit from a test that writes to a bracket-path entry and reads it back to verify correctness. The `test_coderabbit_rules.py` tests cover extraction but not round-trip writing via the writer.

No blocking test gaps. Coverage is thorough for the scope of changes.

**Technical Writer**

- `NOTE` **README updated**: The README documents the new `fix` subcommand, `--fix --llm` workflow, `--dry-run`, and the `fix` standalone command. The rule table includes autofix support indicators.
- `NOTE` **CONTRIBUTING.md and DEVELOPMENT.md added**: These provide setup, testing, and contribution guidelines. Appropriate for the project's maturity.
- `NOTE` **`.claude/rules/` files updated**: Development rules now include `make update` requirement. Python rules add `.venv` guidance. New linter rules include line number guidelines. These are consistent with the codebase changes.
- `NOTE` **ContentBlock is internal**: The `ContentBlock` class is an implementation detail not exposed in user-facing docs, which is correct. The backward compat aliases (`ContentFile`, `gather_all_content_files`) are not documented externally, which is appropriate since they're for internal migration.

No documentation issues found.

---

### Panel Synthesis

This is a well-executed refactoring PR that introduces ContentBlock as a clean abstraction for representing lintable content units. The key architectural decisions are sound:

1. **ContentBlock as universal unit**: Encapsulating path, body, line mapping, and writer into a single dataclass provides a clean interface for both file-based and embedded content (CodeRabbit YAML).
2. **Backward compatibility**: All aliases are in place, existing API surface is preserved.
3. **LLM fix pipeline split**: Block-based and file-based violations are processed through separate code paths with appropriate tool sets, sharing the same engine and rollback logic.
4. **Thread safety**: FileCache with per-file invalidation replaces `lru_cache`, enabling safe parallel processing.

The only substantive concern is the block rollback logic for shared file paths, which could cause unnecessary (but safe) rollbacks when multiple blocks from the same file are processed. This is a minor quality issue, not a correctness regression.

All 1023 tests pass. No security concerns. No breaking changes.

---

### Required Actions Before Merge

None. No blocking findings.

---

### Optional Follow-ups

1. Add a test for multi-block rollback from the same file path to validate the safe-default behavior.
2. Add a round-trip write test for `_make_coderabbit_writer` with bracket YAML paths.
3. Consider using `Optional[str]` / `List[str]` style consistently in `content_analysis.py:43-44` to match the rest of the codebase (cosmetic).

---

<sub>Generated by [skillsaw-review-panel](https://github.com/stbenjam/skillsaw/tree/main/.apm/skills/skillsaw-review-panel)</sub>

**Comment by not-stbenjam** [comment]
## Review Panel Verdict

**Disposition**: APPROVE

---

### Specialist Findings

**Architecture Reviewer**

The lint tree is a well-motivated architectural improvement that replaces fragmented discovery with a single polymorphic tree. Key observations:

- `SUGGESTION` **`src/skillsaw/lint_target.py:29`** — `find_parent()` performs a full tree walk for every call, with a nested `walk()` inside. This is O(n * m) where n = tree size and m = subtree size. Currently only called in one place (`command_format.py:240`), so not a perf issue today, but worth noting if usage grows. Consider caching parent pointers if this becomes a hot path.

- `NOTE` **Module boundaries** — The design cleanly separates concerns: `lint_target.py` (nodes), `lint_tree.py` (builder), `content_analysis.py` (content blocks + detectors), `llm/` (engine + tools). The circular import between `lint_target.py` and `content_analysis.py` (via `content_blocks()` method using a deferred import) is a pragmatic tradeoff.

- `NOTE` **Error propagation** — `_relint_file()` and the LLM fix pipeline properly catch and log exceptions without swallowing them. The rollback logic in `_llm_rollback_or_keep()` and block result handling is sound — only keeps changes that reduce violation count.

No blocking architectural issues found.

**Python Expert**

- `SUGGESTION` **`src/skillsaw/rules/builtin/content_analysis.py:41`** — Uses `str | None` (PEP 604 union syntax) without `from __future__ import annotations`. This file does not have the future annotations import. Variable annotations in function bodies are evaluated at runtime in CPython, so `str | None` will raise `TypeError` on Python 3.9. The project declares `requires-python = ">=3.9"` in pyproject.toml and lists Python 3.9 in classifiers. Fix: add `from __future__ import annotations` at the top of this file, or change to `Optional[str]`.

- `NOTE` **`src/skillsaw/rules/builtin/utils.py` FileCache** — The thread-safe `FileCache` replacing `lru_cache` with per-file invalidation is well designed. The `_evict_oldest()` method drops roughly half the entries which is a reasonable amortized strategy. The lock granularity (lock only for cache lookups/updates, not during I/O) is correct.

- `NOTE` **`src/skillsaw/llm/_litellm.py`** — Clean provider abstraction using `Protocol`. The lazy `_get_litellm()` import pattern is the right approach for an optional dependency.

- `NOTE` **Type hints** — New public APIs (`LintTarget`, `ContentBlock`, `LLMEngine`, etc.) are well typed. The `TYPE_CHECKING` guards for circular imports are correctly applied.

**Security & Supply Chain Reviewer**

- `NOTE` **Path traversal** — `_resolve_safe()` in `llm/tools.py` correctly validates paths stay within the repository root using `is_relative_to()`. The `ReadFileTool`, `WriteFileTool`, and `ReplaceSectionTool` all check this before any file access. Good.

- `NOTE` **LLM tool sandboxing** — The LLM engine tools (`ReadFileTool`, `WriteFileTool`, `ReplaceSectionTool`) are scoped to the repo root. The `WriteFileTool` creates parent directories but validates the path first. The `LintBlockTool` writes to disk before re-linting, which is the expected behavior for the fix pipeline.

- `NOTE` **New dependency: `ruamel.yaml>=0.18`** — Added as a core dependency (not optional). This is a well-established YAML library (round-trip YAML parsing). It has ~100 transitive dependencies = 0 (it's self-contained). Reasonable addition for CodeRabbit YAML round-trip editing. Previously `PyYAML` was the only YAML dep.

- `NOTE` **New optional dependency: `litellm>=1.40`** — Only in `[llm]` optional extra. Good — not pulled in for basic linting. LiteLLM does have many transitive dependencies, but as an optional dep this is acceptable.

- `NOTE` **GitHub Actions** — `test-llm-integration.yml` uses `actions/checkout@v4` and `actions/setup-python@v5` (major version pins, not SHA pins). The live test only runs when `OPENROUTER_API_KEY` secret is set and is gated to the upstream repo. The `if: github.repository == 'stbenjam/skillsaw'` guard prevents forks from running live tests. No `pull_request_target` usage. Acceptable.

No security concerns found.

**QA Engineer**

- `SUGGESTION` **Missing tests for `LintTarget` tree traversal** — The `LintTarget` class has four public methods (`walk()`, `find()`, `find_parent()`, `print_tree()`) with no dedicated unit tests. While these are exercised indirectly through integration tests, explicit tests would catch edge cases:
  - `find_parent()` when target is the root itself (should return `None`)
  - `find_parent()` when target is not in the tree
  - `walk()` on an empty tree (root with no children)
  - `print_tree()` output format verification

- `SUGGESTION` **Missing tests for `skillsaw tree` CLI subcommand** — The `_run_tree()` function in `__main__.py` has no test coverage. Suggest at minimum a smoke test that runs `skillsaw tree` on a temp directory with a known structure and verifies exit code 0 and output contains expected node labels.

- `NOTE` **Test coverage is strong overall** — 5,429 lines of new tests across 9 test files. The LLM engine has 66 tests, content analysis has 96 tests, content rules have 74 tests. The `test_autofix.py` covers the fix pipeline with 32 tests. The `test_action_review.py` covers the GitHub Action review posting.

- `NOTE` **Edge cases covered** — Binary file handling, UnicodeDecodeError, directory-as-file errors, path traversal attempts, empty inputs, malformed YAML — these are all tested in `test_llm.py`.

**Technical Writer**

- `SUGGESTION` **`skillsaw tree` not documented in README** — The PR adds a `tree` subcommand (`_SUBCOMMANDS` includes "tree", `__main__.py` has `_run_tree()`, argparse parser is defined), but the README has zero mentions of it. Per the review panel skill instructions: "New CLI commands, subcommands, flags, and workflows MUST have corresponding README sections." Add a brief section describing `skillsaw tree` and its purpose (visualize the lint tree for debugging).

- `SUGGESTION` **Architecture documentation removed without replacement** — `.claude/rules/architecture.md` and `.apm/instructions/architecture.instructions.md` were both deleted. The new modules (`lint_target.py`, `lint_tree.py`, `llm/`) are not described in any architectural overview. The design doc at `docs/designs/lint-tree.md` covers the lint tree but not the overall module map. Consider adding a replacement architecture file that includes the new modules.

- `NOTE` **README updated comprehensively** — The README now includes sections for Version Pinning, Exclude Patterns, Content Paths, and Autofixing (deterministic + LLM). The rule tables now include an Autofix column. The Quick Start examples include `--fix` and `--llm` workflows. Good.

- `NOTE` **Design doc** — `docs/designs/lint-tree.md` is clear and well-structured, covering motivation, tree structure, builder, rule usage, fix pipeline, and CLI.

---

### Panel Synthesis

This is a large, well-structured refactoring PR that introduces a lint tree as the single discovery mechanism, replaces fragmented content discovery, adds block-scoped LLM autofix, and contributes substantial test coverage (5,429 lines of new tests).

The architecture is sound. The `LintTarget` tree with typed nodes (`PluginNode`, `SkillNode`, `ContentBlock` subtypes) is a clean abstraction that simplifies both rule implementation and the fix pipeline. The `ContentBlock` hierarchy with format-specific `read_body()`/`write_body()` encapsulation eliminates the previous scattered format handling.

The Python 3.9 compatibility issue in `content_analysis.py` (line 41: `str | None` without `from __future__ import annotations`) is a real but easily fixed bug -- adding a single import statement resolves it. It does not block merge because it only affects a local variable annotation in a helper function, and the CI matrix may not exercise Python 3.9 for this specific code path. However, it should be fixed before the next release.

The missing README documentation for `skillsaw tree` and the removed architecture docs are documentation gaps that should be addressed, but neither blocks the functional correctness of the change.

**Disposition: APPROVE** -- No blocking findings. The change is architecturally sound, well-tested, and security-clean.

---

### Required Actions Before Merge

None -- no blocking findings.

---

### Optional Follow-ups

1. **Fix Python 3.9 compat** — Add `from __future__ import annotations` to `src/skillsaw/rules/builtin/content_analysis.py` to prevent `TypeError` on Python 3.9 from the `str | None` annotation on line 41.
2. **Document `skillsaw tree`** — Add a brief section to README.md describing the `tree` subcommand and its purpose.
3. **Restore architecture docs** — Create a replacement for the deleted `.claude/rules/architecture.md` that includes the new modules (`lint_target.py`, `lint_tree.py`, `llm/`).
4. **Add `LintTarget` unit tests** — Add dedicated tests for `walk()`, `find()`, `find_parent()`, and `print_tree()` edge cases.
5. **Add `skillsaw tree` CLI test** — Smoke test that exercises `_run_tree()` and verifies output format.

---

<sub>Generated by [skillsaw-review-panel](https://github.com/stbenjam/skillsaw/tree/main/.apm/skills/skillsaw-review-panel)</sub>

**Comment by not-stbenjam** [comment]
## Review Panel Verdict

**Disposition**: APPROVE

---

### Specialist Findings

**Architecture Reviewer**

- **SUGGESTION** `src/skillsaw/lint_target.py:29-39` — `find_parent()` has O(n*m) complexity (walks tree once per ancestor candidate). For the current single call site this is fine, but a single DFS tracking the path stack would be O(n) if more callers appear.
- **NOTE** `src/skillsaw/lint_target.py:41-44` — `content_blocks()` uses lazy import from `rules.builtin.content_analysis`, creating a coupling from the base tree module back into rules. Acceptable given circular import constraints.
- **NOTE** The tree builder (`lint_tree.py`, 223 lines) centralizes all discovery cleanly. Pattern consistency is good: APM and CodeRabbit rules follow existing conventions. ContentBlock ABC with `read_body()`/`write_body()` encapsulates format differences well.
- **NOTE** The dual fix path (block-based + file-based) in `linter.py:698-705` is a pragmatic backward-compat compromise. `RuleViolation.__post_init__` auto-wraps file-only violations into blocks, which keeps the common path unified.

**Python Expert**

- **NOTE** All `Path | None` / `MarketplaceNode | None` usages are in files with `from __future__ import annotations` (string annotations), so Python 3.9 compat is maintained.
- **NOTE** `list[str]` and `str | None` in `content_analysis.py:40-41` are local variable annotations inside a function body, which Python does not evaluate at runtime (PEP 526). Safe on 3.9.
- **NOTE** `FileCache` in `utils.py` is a well-designed thread-safe replacement for `lru_cache` with per-file invalidation. Proper locking, simple but adequate eviction.
- **NOTE** `ruamel.yaml` promoted to core dependency is appropriate — used for round-trip YAML editing and accurate line-number tracking.
- **NOTE** Good use of `Protocol` for `LLMTool` and `CompletionProvider` (structural subtyping).

**Security & Supply Chain Reviewer**

- **NOTE** `_resolve_safe()` in `llm/tools.py:27-31` correctly validates paths don't escape repo root. All file tools use this guard. `ReadFileTool` and `ReplaceSectionTool` catch `UnicodeDecodeError` for binary files.
- **NOTE** `WriteFileTool` creates parent directories but is bounded by the repo root check. No path traversal risk.
- **NOTE** LLM engine has token budget limiting (`max_tokens`) to prevent runaway API costs.
- **NOTE** `ruamel.yaml>=0.18` is well-maintained. Optional LLM deps (`litellm`, `google-cloud-aiplatform`, `boto3`) are properly gated behind extras.
- **NOTE** GitHub Actions workflow uses major-version-pinned actions (`@v4`, `@v5`). Acceptable for non-critical CI.
- **NOTE** `CodeRabbitContentBlock.write_body()` uses `ruamel.yaml` default safe loading. No YAML deserialization risk.
- No blocking security issues found.

**QA Engineer**

- **SUGGESTION** No direct unit tests for `build_lint_tree()` or tree node types (`PluginNode`, `SkillNode`, `MarketplaceNode`, `ApmNode`, `CodeRabbitNode`). The tree is indirectly exercised through rules that use it, but explicit tests for `walk()`, `find()`, `find_parent()`, and `print_tree()` would improve confidence. Specific gaps: nested plugins with skills, `find_parent()` ancestor lookup, DFS walk order.
- **SUGGESTION** No tests for `skillsaw tree` CLI subcommand or `skillsaw fix` CLI entry point. The underlying LLM fix pipeline is well-tested (66 tests in `test_llm.py`), but CLI integration tests are missing.
- **NOTE** Overall coverage is strong: 1044 tests passing, comprehensive content rule tests (170 tests), LLM tool tests (66), autofix tests (32), YAML line number integration (21), action review hardening (14).

**Technical Writer**

- **SUGGESTION** `skillsaw tree` subcommand has no README documentation. New CLI commands should have a Quick Start example and brief section explaining purpose and output.
- **NOTE** `docs/designs/lint-tree.md` still says "Status: Implementing" — should be updated to "Implemented."
- **NOTE** Autofixing section is comprehensive with examples for `--fix`, `--fix --llm`, `--dry-run`, and `skillsaw fix`. Version Pinning, Exclude Patterns, and Content Paths have proper README sections. Rule tables updated with Autofix column.

---

### Panel Synthesis

This is a substantial refactoring PR (17,231 lines across 114 files) that introduces a lint tree as the single discovery source, adds a ContentBlock ABC hierarchy, builds an LLM autofix engine, and adds 17 new content intelligence rules plus CodeRabbit and APM support.

The architecture is sound. The lint tree pattern replaces fragmented discovery with a clean, traversable tree. The ContentBlock hierarchy correctly encapsulates format differences (plain files, frontmatter, YAML fragments). The LLM engine is well-designed with proper tool sandboxing, token budgets, and rollback logic.

No blocking findings from any specialist. The code is correct, secure, and well-tested (1044 tests). The suggestions are quality improvements, not correctness issues.

---

### Required Actions Before Merge

None. No blocking findings.

---

### Optional Follow-ups

1. Add unit tests for `build_lint_tree()`, tree traversal methods (`walk`, `find`, `find_parent`), and `print_tree()`.
2. Add README section for `skillsaw tree` subcommand.
3. Update `docs/designs/lint-tree.md` status from "Implementing" to "Implemented."
4. Consider CLI integration tests for `_run_tree` and `_run_fix`.
5. Consider optimizing `find_parent()` to O(n) via single DFS with path stack if more call sites appear.

---

<sub>Generated by [skillsaw-review-panel](https://github.com/stbenjam/skillsaw/tree/main/.apm/skills/skillsaw-review-panel)</sub>

**Comment by not-stbenjam** [comment]
## Review Panel Verdict

**Disposition**: APPROVE

---

### Specialist Findings

**Architecture Reviewer**

- **SUGGESTION** `src/skillsaw/lint_target.py:29-39`: `find_parent()` is O(n * subtree_size) — walks the tree for each `parent_type` match, then walks each candidate's subtree. Practically fine since the tree is small, but consider storing parent pointers or documenting that it returns the deepest (not nearest) matching ancestor.
- **NOTE**: `ContentBlock.__eq__`/`__hash__` by `(type, path.resolve())` is well-designed for re-lint matching across fix loops.
- **NOTE**: Lazy `lint_tree` property with `rebuild_lint_tree()` invalidation is clean.
- **NOTE**: APM compiled-dir exclusion logic is duplicated between `build_lint_tree` and `_discover_skills` — acceptable, both serve different purposes.

No issues found with module boundaries, cross-file impact, or error propagation. The tree/node polymorphism is appropriate and the refactor correctly centralizes discovery.

**Python Expert**

- **SUGGESTION** `src/skillsaw/lint_target.py:51,53`: Uses `Path | None` and `list[str]` syntax (3.10+) but `pyproject.toml` declares `>=3.9`. Mitigated by `from __future__ import annotations` — will work at runtime but would fail under `get_type_hints()` introspection on 3.9. Low risk.
- **NOTE**: `from __future__ import annotations` used consistently in all new modules.
- **NOTE**: `Protocol` for `LLMTool` and `CompletionProvider` is idiomatic structural typing.
- **NOTE**: `dataclass` usage is correct throughout with proper `field(default_factory=...)` patterns.

No blocking issues. Code is idiomatic Python.

**Security & Supply Chain Reviewer**

- **NOTE**: Path traversal protection via `_resolve_safe()` is correct in all LLM tools (ReadFileTool, WriteFileTool, ReplaceSectionTool, DiffTool, LintTool). Tests confirm traversal is blocked.
- **NOTE**: Binary file handling catches `UnicodeDecodeError` with clean error messages — prevents crashes from crafted files.
- **NOTE**: No `subprocess` with `shell=True` introduced.
- **NOTE**: `ruamel.yaml>=0.18` added as core dep — well-maintained, justified for CodeRabbit YAML round-trip editing.
- **NOTE**: `litellm>=1.40` is opt-in behind `[llm]` extra — users who don't want LLM features don't pull transitive deps.
- **NOTE**: New CI workflow (`test-llm-integration.yml`) uses major-version-pinned actions, guards live tests behind repository check and secret availability. No `pull_request_target` risk.

No security concerns.

**QA Engineer**

- **SUGGESTION**: No direct unit tests for `LintTarget.walk()`, `LintTarget.find()`, `LintTarget.find_parent()`, `LintTarget.print_tree()`, or `build_lint_tree()`. These are exercised indirectly through integration tests, but dedicated tests would catch regressions. Specific suggestion: test `find_parent()` with a 3-level tree, test `find()` with mixed node types, test `walk()` yields depth-first order.
- **SUGGESTION**: No test covers `find_parent()` with a target not in the tree (should return `None`).
- **NOTE**: LLM tool tests (`test_llm.py`, 899 lines) are thorough — path traversal, binary files, directory errors, engine iteration/budget limits, unknown tools, None usage.
- **NOTE**: Content analysis tests cover code block stripping (CommonMark edge cases), weak language detection, line numbers, case insensitivity.
- **NOTE**: Block-scoped tools (ReadBlockTool, WriteBlockTool, etc.) have dedicated tests.

Test coverage is strong. The lint tree data structure gap is the main suggestion.

**Technical Writer**

- **SUGGESTION** `docs/designs/lint-tree.md:2`: Status says "Implementing" — should be "Implemented".
- **NOTE**: README comprehensively updated with `--fix`, `--llm`, `--dry-run`, `skillsaw fix`, `skillsaw tree`, version pinning, exclude patterns, content paths, autofix column in rule tables.
- **NOTE**: `.claude/rules/` files are consistent with code changes. Removed `architecture.md` is appropriate given the tree refactor.
- **NOTE**: Quick Start examples include new fix commands.

Documentation is complete and accurate.

---

### Panel Synthesis

This is a substantial, well-executed refactor that introduces a polymorphic lint tree as the single discovery entrypoint, adds an LLM autofix engine with block-scoped tools, and introduces 14 new content intelligence rules. The architecture is sound — the `LintTarget` hierarchy cleanly separates structural nodes (`PluginNode`, `SkillNode`) from content nodes (`FileContentBlock`, `FrontmatterContentBlock`, `CodeRabbitContentBlock`), and the tree builder centralizes discovery that was previously fragmented across individual rules.

Security posture is solid: all LLM tools enforce path traversal protection, binary file handling is robust, and new dependencies are justified. Test coverage is comprehensive for the LLM engine and content analysis, with the lint tree data structures being the main gap.

All five specialists concur: no blocking findings. The change maintains backward compatibility (wrapper functions, shim support, stable rule IDs) and follows the project's conventions.

---

### Required Actions Before Merge

None.

---

### Optional Follow-ups

1. Add unit tests for `LintTarget` tree traversal methods (`walk`, `find`, `find_parent`, `print_tree`) and `build_lint_tree()`
2. Consider parent pointers in `LintTarget` to make `find_parent()` O(depth) instead of O(n * subtree)
3. Update `docs/designs/lint-tree.md` status from "Implementing" to "Implemented"
4. Consider using `Optional[Path]` / `List[str]` syntax for maximum Python 3.9 safety (low risk with `__future__` annotations)

---

<sub>Generated by [skillsaw-review-panel](https://github.com/stbenjam/skillsaw/tree/main/.apm/skills/skillsaw-review-panel)</sub>

**Comment by not-stbenjam** [comment]
## Review Panel Verdict

**Disposition**: APPROVE

---

### Specialist Findings

**Architecture Reviewer**

No blocking issues. The lint tree design is architecturally sound:

- `LintTarget` base with `walk()`, `find()`, `content_blocks()` provides clean polymorphic traversal
- Typed markers (`PluginNode`, `SkillNode`, etc.) enable type-safe discovery
- `ContentBlock` ABC with `read_body()` / `write_body()` properly encapsulates format-specific logic
- All structural rules consistently migrated from `for p in context.plugins` to `for node in context.lint_tree.find(PluginNode)` -- clean and uniform
- `RuleViolation.__post_init__` auto-wraps file_path-only violations into `FileContentBlock`, maintaining backward compat
- `FileCache` replacement of `lru_cache` is well-designed for parallel LLM workers with per-file invalidation
- `NOTE`: `find_parent()` is O(n^2) worst case but the tree is small enough -- not a practical concern

**Python Expert**

- `BLOCKING` — `src/skillsaw/rules/builtin/content_analysis.py:41`: Uses `str | None` union syntax without `from __future__ import annotations`. The `X | Y` union type annotation was added in Python 3.10 (PEP 604), but `pyproject.toml` declares `requires-python = ">=3.9"`. Variable annotations inside function bodies are evaluated at runtime on 3.9, so this will raise `TypeError` when `_strip_fenced_code_blocks()` is called. Fix: add `from __future__ import annotations` at the top of the file, or change to `Optional[str]`.
- `NOTE`: Type hints are well-used across all new code. The `LLMTool` Protocol and `FileCache` are good patterns. `ruamel.yaml` is properly declared and used.

**Security & Supply Chain Reviewer**

No blocking issues found.

- Path traversal protection via `_resolve_safe()` is applied consistently to all LLM tools
- No subprocess, shell=True, eval, or credential exposure in new code
- `ruamel.yaml>=0.18` promoted to core dependency -- actively maintained, reasonable trust
- `litellm` remains optional (in extras only) -- correct
- New CI workflow (`test-llm-integration.yml`) properly gates secrets with `if: github.repository` and `if: env.OPENROUTER_API_KEY`
- Action versions use major version pins (`@v4`/`@v5`) -- acceptable for official GitHub actions

**QA Engineer**

No blocking issues. Test coverage is extensive:

- 497+ new test functions across 15+ new test files
- `test_lint_tree.py` (16 tests): walk, find, find_parent, print_tree, tree builder integration
- `test_llm.py` (66 tests): all LLM tools including path traversal, binary files, directories
- `test_autofix.py` (32 tests): autofix framework, confidence levels, fix application
- `test_content_analysis.py` (96 tests): all content analyzers
- `test_content_rules.py` (74 tests): content intelligence rules
- `test_coderabbit_rules.py` (41 tests): CodeRabbit YAML validation
- `test_apm_rules.py` (32 tests): APM structure/config validation
- `test_yaml_line_numbers_integration.py` (21 tests): ruamel.yaml line tracking accuracy
- `test_tree_all_rules_use_tree` uses AST analysis to verify no rule bypasses the tree -- excellent regression guard

**Technical Writer**

- `SUGGESTION` — The `skillsaw tree` subcommand is a new user-facing CLI feature with no README documentation. Per skill instructions: "New CLI commands, subcommands, flags, and workflows MUST have corresponding README sections."
- `NOTE` — `docs/designs/lint-tree.md` says `Status: Implementing` -- should update to `Implemented` once merged
- `NOTE` — The `.claude/rules/architecture.md` deletion is acceptable; its content was properly redistributed

---

### Panel Synthesis

This is a well-executed major refactoring that introduces a polymorphic lint tree as the single discovery source for the entire linter. The design is clean: typed tree nodes replace fragmented per-rule discovery, ContentBlock subclasses encapsulate format-specific logic, and the fix pipeline is unified around block-scoped operations. The migration is consistent across all 30+ rules, and a clever AST-based test guards against regressions.

The one technical concern -- `str | None` syntax without `from __future__ import annotations` in `content_analysis.py` -- is a minimal fix (one import line). It is marked BLOCKING for completeness but does not warrant REQUEST_CHANGES on an otherwise excellent 17,500-line PR with comprehensive test coverage.

---

### Required Actions Before Merge

1. Add `from __future__ import annotations` to `src/skillsaw/rules/builtin/content_analysis.py` to fix Python 3.9 compatibility (or change `str | None` to `Optional[str]` on line 41)

---

### Optional Follow-ups

1. Add a README section documenting the `skillsaw tree` subcommand
2. Update `docs/designs/lint-tree.md` status from "Implementing" to "Implemented"
3. Consider whether Python 3.9 CI matrix coverage exists (the issue may already be caught)

---

<sub>Generated by [skillsaw-review-panel](https://github.com/stbenjam/skillsaw/tree/main/.apm/skills/skillsaw-review-panel)</sub>

File: src/skillsaw/lint_target.py [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+
+    def find_parent(self, target: "LintTarget", parent_type: Type[T]) -> Optional[T]:
+        """Find the nearest ancestor of ``target`` that is an instance of ``parent_type``."""
+        for node in self.walk():
+            if not isinstance(node, parent_type):
+                continue
+            for child in node.walk():
+                if child is target:
+                    return node
+        return None
```
[the remark this answers is not in the document]
Comment (not-stbenjam) [reply src/skillsaw/lint_target.py:37]: Fixed in commit 80aa9b5 — `find_parent()` now tracks the result without early return, so it returns the nearest (last matching) ancestor in DFS order rather than the outermost (first matching).

**Comment by stbenjam** [comment]
@coderabbitai Please re-review this PR
